File: packages/typescript/service/src/aggregation.ts
/**
 * Audit aggregation — collapses repeated low-value denials into summary records.
 */

export interface AggregatedEntry {
  event_class: "repeated_low_value_denial";
  retention_tier: "aggregate_only";
  grouping_key: { actor_key: string; capability: string; failure_type: string };
  window_start: string;
  window_end: string;
  count: number;
  first_seen: string;
  last_seen: string;
  representative_detail: string | null;
  toAuditDict(): Record<string, unknown>;
}

function bucketKey(timestampIso: string, windowSeconds: number): number {
  const epoch = Math.floor(new Date(timestampIso).getTime() / 1000);
  return epoch - (epoch % windowSeconds);
}

function groupingKey(event: Record<string, unknown>): string {
  const ak = String(event.actor_key ?? "anonymous");
  const cap = String(event.capability ?? "_pre_auth");
  const ft = String(event.failure_type ?? "unknown");
  return `${ak}\0${cap}\0${ft}`;
}

function parseGroupingKey(key: string): { actor_key: string; capability: string; failure_type: string } {
  const [actor_key, capability, failure_type] = key.split("\0");
  return { actor_key, capability, failure_type };
}

function createAggregatedEntry(
  gk: { actor_key: string; capability: string; failure_type: string },
  events: Record<string, unknown>[],
  bucketEpoch: number,
  windowSeconds: number,
): AggregatedEntry {
  const timestamps = events.map((e) => String(e.timestamp));
  const firstDetail = events[0].detail;
  let repDetail: string | null = null;
  if (firstDetail != null) {
    repDetail = String(firstDetail).slice(0, 200);
  }

  const windowStart = new Date(bucketEpoch * 1000).toISOString();
  const windowEnd = new Date((bucketEpoch + windowSeconds) * 1000).toISOString();

  return {
    event_class: "repeated_low_value_denial",
    retention_tier: "aggregate_only",
    grouping_key: gk,
    window_start: windowStart,
    window_end: windowEnd,
    count: events.length,
    first_seen: timestamps.sort()[0],
    last_seen: timestamps.sort()[timestamps.length - 1],
    representative_detail: repDetail,
    toAuditDict() {
      return {
        entry_type: "aggregated",
        event_class: this.event_class,
        retention_tier: this.retention_tier,
        grouping_key: this.grouping_key,
        window: { start: this.window_start, end: this.window_end },
        count: this.count,
        first_seen: this.first_seen,
        last_seen: this.last_seen,
        representative_detail: this.representative_detail,
      };
    },
  };
}

export class AuditAggregator {
  private _windowSeconds: number;
  private _buckets: Map<number, Map<string, Record<string, unknown>[]>> = new Map();

  constructor(opts?: { windowSeconds?: number }) {
    this._windowSeconds = opts?.windowSeconds ?? 60;
  }

  submit(event: Record<string, unknown>): void {
    const ts = String(event.timestamp ?? new Date().toISOString());
    const bucket = bucketKey(ts, this._windowSeconds);
    const gk = groupingKey(event);

    if (!this._buckets.has(bucket)) {
      this._buckets.set(bucket, new Map());
    }
    const bucketMap = this._buckets.get(bucket)!;
    if (!bucketMap.has(gk)) {
      bucketMap.set(gk, []);
    }
    bucketMap.get(gk)!.push(event);
  }

  getPendingCount(): number {
    return this._buckets.size;
  }

  flush(now: Date): Array<Record<string, unknown> | AggregatedEntry> {
    const nowEpoch = Math.floor(now.getTime() / 1000);
    const results: Array<Record<string, unknown> | AggregatedEntry> = [];
    const closedBuckets: number[] = [];

    const sortedKeys = [...this._buckets.keys()].sort((a, b) => a - b);
    for (const bucketEpoch of sortedKeys) {
      const bucketEnd = bucketEpoch + this._windowSeconds;
      if (bucketEnd > nowEpoch) continue;

      closedBuckets.push(bucketEpoch);
      const bucketMap = this._buckets.get(bucketEpoch)!;
      for (const [gkStr, events] of bucketMap) {
        if (events.length === 1) {
          results.push(events[0]);
        } else {
          results.push(
            createAggregatedEntry(
              parseGroupingKey(gkStr),
              events,
              bucketEpoch,
              this._windowSeconds,
            ),
          );
        }
      }
    }

    for (const b of closedBuckets) {
      this._buckets.delete(b);
    }
    return results;
  }
}

File: packages/go/service/aggregation.go
package service

import (
	"sync"
	"time"
)

// groupingKey is the tuple used to group events in an aggregation window.
type groupingKey struct {
	ActorKey    string
	Capability  string
	FailureType string
}

// bucket tracks events within a single aggregation window for one grouping key.
type bucket struct {
	events               []map[string]any
	firstSeen            time.Time
	lastSeen             time.Time
	representativeDetail string
}

// AggregatedEntry is emitted when a window closes with >1 event for a grouping key.
type AggregatedEntry struct {
	EventClass           string
	RetentionTier        string
	GroupingKey           map[string]string
	WindowStart          string
	WindowEnd            string
	Count                int
	FirstSeen            string
	LastSeen             string
	RepresentativeDetail string
}

// ToAuditDict converts an AggregatedEntry to a map suitable for audit persistence.
func (ae *AggregatedEntry) ToAuditDict() map[string]any {
	return map[string]any{
		"entry_type":            "aggregated",
		"event_class":           ae.EventClass,
		"retention_tier":        ae.RetentionTier,
		"grouping_key":          ae.GroupingKey,
		"aggregation_window":    map[string]string{"start": ae.WindowStart, "end": ae.WindowEnd},
		"aggregation_count":     ae.Count,
		"count":                 ae.Count,
		"first_seen":            ae.FirstSeen,
		"last_seen":             ae.LastSeen,
		"representative_detail": ae.RepresentativeDetail,
		"capability":            ae.GroupingKey["capability"],
		"failure_type":          ae.GroupingKey["failure_type"],
		"success":               false,
	}
}

// windowKey combines a grouping key with a window epoch for map indexing.
type windowKey struct {
	gk    groupingKey
	epoch int64
}

// AuditAggregator buckets low-value denial events by time window.
// Implements SPEC §6.9.
type AuditAggregator struct {
	windowSeconds int64
	mu            sync.Mutex
	windows       map[windowKey]*bucket
}

// NewAuditAggregator creates an aggregator with the given window size in seconds.
func NewAuditAggregator(windowSeconds int) *AuditAggregator {
	return &AuditAggregator{
		windowSeconds: int64(windowSeconds),
		windows:       make(map[windowKey]*bucket),
	}
}

// Submit adds an event to the aggregator.
func (a *AuditAggregator) Submit(event map[string]any) {
	a.mu.Lock()
	defer a.mu.Unlock()

	actorKey, _ := event["actor_key"].(string)
	if actorKey == "" {
		actorKey = "anonymous"
	}
	capability, _ := event["capability"].(string)
	if capability == "" {
		capability = "_pre_auth"
	}
	failureType, _ := event["failure_type"].(string)
	if failureType == "" {
		failureType = "unknown"
	}

	ts := parseTimestamp(event)
	epoch := ts.Unix() - (ts.Unix() % a.windowSeconds)

	gk := groupingKey{ActorKey: actorKey, Capability: capability, FailureType: failureType}
	wk := windowKey{gk: gk, epoch: epoch}

	b, ok := a.windows[wk]
	if !ok {
		detail, _ := event["detail"].(string)
		if len(detail) > 200 {
			detail = detail[:200]
		}
		b = &bucket{
			firstSeen:            ts,
			lastSeen:             ts,
			representativeDetail: detail,
		}
		a.windows[wk] = b
	}
	b.events = append(b.events, event)
	if ts.After(b.lastSeen) {
		b.lastSeen = ts
	}
}

// Flush closes all windows whose end time <= now and returns the results.
// Single-event buckets pass through as map[string]any.
// Multi-event buckets produce *AggregatedEntry.
func (a *AuditAggregator) Flush(now time.Time) []any {
	a.mu.Lock()
	defer a.mu.Unlock()

	var results []any
	nowUnix := now.Unix()

	for wk, b := range a.windows {
		windowEnd := wk.epoch + a.windowSeconds
		if windowEnd > nowUnix {
			continue // window still open
		}

		windowStartStr := time.Unix(wk.epoch, 0).UTC().Format(time.RFC3339)
		windowEndStr := time.Unix(windowEnd, 0).UTC().Format(time.RFC3339)

		if len(b.events) == 1 {
			results = append(results, b.events[0])
		} else {
			results = append(results, &AggregatedEntry{
				EventClass:    "repeated_low_value_denial",
				RetentionTier: "aggregate_only",
				GroupingKey: map[string]string{
					"actor_key":    wk.gk.ActorKey,
					"capability":   wk.gk.Capability,
					"failure_type": wk.gk.FailureType,
				},
				WindowStart:          windowStartStr,
				WindowEnd:            windowEndStr,
				Count:                len(b.events),
				FirstSeen:            b.firstSeen.UTC().Format(time.RFC3339),
				LastSeen:             b.lastSeen.UTC().Format(time.RFC3339),
				RepresentativeDetail: b.representativeDetail,
			})
		}
		delete(a.windows, wk)
	}

	return results
}

func parseTimestamp(event map[string]any) time.Time {
	if ts, ok := event["timestamp"].(string); ok {
		if t, err := time.Parse(time.RFC3339, ts); err == nil {
			return t
		}
	}
	return time.Now().UTC()
}
